src/jit/lowerxarch.cpp

```diff
@@ -805,6 +805,77 @@ void Lowering::LowerHWIntrinsicCC(GenTreeHWIntrinsic* node, NamedIntrinsic newIn
     }
 }
 
+//----------------------------------------------------------------------------------------------
+// LowerFusedMultiplyAdd: Changes NI_FMA_MultiplyAddScalar produced by Math(F).FusedMultiplyAdd
+//     to a better FMA intrinsics if there are GT_NEG around in order to eliminate them.
+//
+//  Arguments:
+//     node - The hardware intrinsic node
+//
+//  Notes:
+//     Math(F).FusedMultiplyAdd is expanded into NI_FMA_MultiplyAddScalar and
+//     depending on additional GT_NEG nodes around it can be:
+//
+//      x *  y + z -> NI_FMA_MultiplyAddScalar
+//      x * -y + z -> NI_FMA_MultiplyAddNegatedScalar
+//     -x *  y + z -> NI_FMA_MultiplyAddNegatedScalar
+//     -x * -y + z -> NI_FMA_MultiplyAddScalar
+//      x *  y - z -> NI_FMA_MultiplySubtractScalar
+//      x * -y - z -> NI_FMA_MultiplySubtractNegatedScalar
+//     -x *  y - z -> NI_FMA_MultiplySubtractNegatedScalar
+//     -x * -y - z -> NI_FMA_MultiplySubtractScalar
+//
+void Lowering::LowerFusedMultiplyAdd(GenTreeHWIntrinsic* node)
+{
+    GenTreeArgList*     argList = node->gtGetOp1()->AsArgList();
+    GenTreeHWIntrinsic* createScalarOps[4]; // 3 arguments with 1-based index
+
+    for (int i = 1; i < 4; i++)
+    {
+        assert(argList->Current() != nullptr);
+        if (argList->Current()->OperIsHWIntrinsic())
+        {
+            GenTreeHWIntrinsic* hwArg = argList->Current()->AsHWIntrinsic();
```

Comment:
We can't assume that the args are HWIntrinsic nodes. They could be directly consumed local vars or indirections.

e.g. the following is completely valid and should violate this assumption:
```csharp
void MyFunc(Vector128<float>* pInOut)
{
    Vector128<float> x = Vector128.CreateScalar(5.0f);
    *pInOut = Fma.MultiplyAddScalar(*pInOut, x, x);
}
```

Reply:
@tannergooding the goal was to improve `Math(F).FusedMutlitplyAdd`. When developers use intrinsics API directly they probably know what they are doing. So should I make the code more complicated to handle all possible cases?